metricflow/dataflow/builder/builder_cache.py

```diff
@@ -0,0 +1,78 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Optional
+
+from metricflow_semantics.collection_helpers.lru_cache import LruCache
+from metricflow_semantics.specs.linkable_spec_set import LinkableSpecSet
+from metricflow_semantics.specs.metric_spec import MetricSpec
+from metricflow_semantics.specs.where_filter.where_filter_transform import WhereSpecFactory
+
+from metricflow.dataflow.builder.measure_spec_properties import MeasureSpecProperties
+from metricflow.dataflow.builder.source_node_recipe import SourceNodeRecipe
+from metricflow.dataflow.dataflow_plan import DataflowPlanNode
+from metricflow.plan_conversion.node_processor import PredicatePushdownState
+
+
+@dataclass(frozen=True)
+class FindSourceNodeRecipeParameterSet:
+    """Parameters for `DataflowPlanBuilder._find_source_node_recipe()`."""
+
+    linkable_spec_set: LinkableSpecSet
+    predicate_pushdown_state: PredicatePushdownState
+    measure_spec_properties: Optional[MeasureSpecProperties]
+
+
+@dataclass(frozen=True)
+class FindSourceNodeRecipeResult:
+    """Result for `DataflowPlanBuilder._find_source_node_recipe()`."""
+
+    source_node_recipe: Optional[SourceNodeRecipe]
+
+
+@dataclass(frozen=True)
+class BuildAnyMetricOutputNodeParameterSet:
+    """Parameters for `DataflowPlanBuilder._build_any_metric_output_node()`."""
+
+    metric_spec: MetricSpec
+    queried_linkable_specs: LinkableSpecSet
+    filter_spec_factory: WhereSpecFactory
+    predicate_pushdown_state: PredicatePushdownState
+    for_group_by_source_node: bool
+
+
+class DataflowPlanBuilderCache:
+    """Cache for internal methods in `DataflowPlanBuilder`."""
+
+    def __init__(  # noqa: D107
+        self, find_source_node_recipe_cache_size: int = 1000, build_any_metric_output_node_cache_size: int = 1000
+    ) -> None:
+        self._find_source_node_recipe_cache = LruCache[FindSourceNodeRecipeParameterSet, FindSourceNodeRecipeResult](
```

Comment:
It would be pretty cool if we could use the new `LruCache` while still using decorator syntax to maintain readability of the methods that are cached. But I guess it's also debatable if that's more readable (e.g., it obscures the cache code if you end up needing to debug something there). And probably not worth the effort it would take to make it work anyway...

Reply:
Yeah, the obscuring of the dependencies with a decorator would make it harder to follow.